src/Reflection/InitializerExprTypeResolver.php

```diff
@@ -440,6 +440,16 @@ public function getConcatType(Expr $left, Expr $right, callable $getTypeCallback
 	 */
 	public function getArrayType(Expr\Array_ $expr, callable $getTypeCallback): Type
 	{
+		// degrade oversized constant arrays, supports only basic types. less precise but slim and fast.
+		if (count($expr->items) > ConstantArrayTypeBuilder::ARRAY_COUNT_LIMIT) {
```

Comment:
1) You need to cache the result in the `$expr` attributes
2) You don't need to reimplement `Expr -> Type` again, you need to take advantage of the fact you're inside InitializerExprTypeResolver already and you have `$getTypeCallback` at your disposal.
3) No need to write a new class - OversizedConstantArrayResolver, it can be just a private method here.
4) You're not gonna need to refactor and create GetType helper either.

Reply:
thank you. Fixed 2,3,4.

why do we need the caching within the attributes?